File: Lab9/Sea.py
# Файл: Sea.py
from WaterBody import WaterBody

class Sea(WaterBody):
    """
    Клас 'Море', що успадковує клас 'Водойма'.
    """
    def __init__(self, name, area, salinity, is_stormy=False):
        """
        Конструктор похідного класу.
        :param name: Назва (str)
        :param area: Площа (float)
        :param salinity: Солоність у проміле (float)
        :param is_stormy: Чи є зараз шторм (bool)
        """
        # Виклик конструктора батьківського класу
        super().__init__(name, area)
        self.salinity = salinity
        self.is_stormy = is_stormy

    def get_info(self):
        """
        Перевизначений метод (Override).
        Додає інформацію про солоність та стан моря.
        """
        base_info = super().get_info()
        storm_status = "Штормить" if self.is_stormy else "Спокійно"
        return f"Тип: Море | {base_info} | Солоність: {self.salinity}‰ | Шторм: {storm_status}"

    def set_storm_status(self, status):
        """Змінює статус шторму."""
        self.is_stormy = status
        print(f"Стан моря '{self.name}' змінено. Шторм: {self.is_stormy}")

File: Lab3/src/KI306/Markovich/Lab3/Sea.java
package KI306.Markovich.Lab3;

import java.io.FileNotFoundException;

/**
 * Клас "Море", що розширює Водойму та реалізує інтерфейс судноплавства.
 * @author Markovich
 */
public class Sea extends WaterBody implements Navigable {
    
    private double salinity; // Солоність у проміле
    private boolean hasStorm;

    /**
     * Конструктор.
     * @param name Назва моря
     * @param len Довжина
     * @param wid Ширина
     * @param depth Глибина
     * @param salinity Солоність
     * @throws FileNotFoundException
     */
    public Sea(String name, double len, double wid, double depth, double salinity) throws FileNotFoundException {
        super(name, len, wid, depth); // Виклик конструктора батька
        this.salinity = salinity;
        this.hasStorm = false;
        log("Створено море. Солоність: " + salinity + " проміле.");
    }

    // Реалізація абстрактного методу
    @Override
    public void analyzeEcosystem() {
        String ecoType = (salinity > 35) ? "Солона (океанічна)" : "Слабосолона";
        System.out.println("Екосистема моря: " + ecoType);
        log("Проведено аналіз екосистеми: " + ecoType);
    }

    // Реалізація методу інтерфейсу
    @Override
    public boolean isNavigationAllowed() {
        boolean allowed = !hasStorm;
        log("Перевірка навігації. Дозволено: " + allowed);
        return allowed;
    }

    // Реалізація методу інтерфейсу
    @Override
    public void declareStormWarning() {
        this.hasStorm = true;
        System.out.println("УВАГА! Штормове попередження на морі " + name);
        log("Оголошено шторм!");
    }
    
    // Власний метод моря
    public void calmDown() {
        this.hasStorm = false;
        System.out.println("Море заспокоїлось.");
        log("Шторм закінчився.");
    }
    
    // Перевизначення методу toString() (або printInfo)
    @Override
    public void printInfo() {
        super.printInfo(); // Виклик методу батька
        System.out.println("Тип: Море | Солоність: " + salinity + "‰ | Шторм: " + hasStorm);
    }
}